### DISCUSSION TITLE
Using RR when cross compiling

### DISCUSSION THREAD

**Discussion by itsLydt** [robotraconteur/robotraconteur#137] [Q&A]
Hello, I am trying to use RR with a Beaglebone Black, but am having some issues. Typically I cross compile on Ubuntu. To this end I have installed robotraconteur-dev on both the target and host, and have synced the files to the target sysroot as well. I am able to compile the example projects on ubuntu for ubuntu, but am struggling to successfully cross compile for beaglebone. The build currently fails when the RobotRaconteurGen inside the beaglebone sysroot is invoked with an exec format error, which makes sense, but if I try to point CMake to the host version of RobotRaconteurGen , I get the error "Could not find a configuration file for package "RobotRaconteur" that is compatible with requested version "".". Is there an example that explains how to set up for cross compilation?

**Comment by johnwason** [comment]
What OS image are you using? Can you run RobotRaconteurGen on the beaglebone installed from robotraconteur-dev?

**Reply by itsLydt** [reply]
The beaglebone is running debian 10 (buster), and while I have not yet tried to compile on the beaglebone directly, I can run `$RobotRaconteurGen --version`  which outputs `RobotRaconteurGen version 0.18.0`

**Comment by johnwason** [comment]
You should be able to set the `RobotRaconteurGen_EXECUTABLE` CMake variable to the host RobotRaconteurGen to avoid the exec format error.

```
cmake -DRobotRaconteurGen_EXECUTABLE=/usr/bin/RobotRaconteurGen ...
```

**Reply by itsLydt** [reply]
Thanks, this seemed to fix the initial error I was getting but still having issues. 
Now the link fails with "cannot find -lz" and I cannot understand why as I have the libz dev package installed in the target sysroot and on the host, and the CMAKE_PREFIX_PATH variable includes the paths of libz.a, libz.so, libz.so.1, etc.

**Reply by itsLydt** [reply]
Looks as though adding `RobotRaconteurCore ` to `target_link_libraries` inCmakeLists.txt adds `-lz` to the `build.ninja` file, and nothing that I put in cmake variables changes it to look for libz in the target sysroot. I'm not super familiar with cmake though, is there a simple way to fix this?

**Reply by johnwason** [reply]
libz is added as a dependency around here https://github.com/robotraconteur/robotraconteur/blob/2b1abdd784c9dfd67038cffce2f45cf8d707ef4e/RobotRaconteurCore/CMakeLists.txt#L306

It does not have a CMake target so it is searching the default compiler library path. Depending on your cross-compiler it should be searching in the sysroot, however the default configuration can vary quite a bit. The library search paths can be modified using LIBRARY_PATH env variable or compiler flags if needed. You can use `gcc -v` to see what the default settings are.

**Reply by itsLydt** [reply]
I think my issue may be due to a cludgy change I made earlier, which I have now reverted. I had changed the file `RobotRaconteurCoreTargets.cmake` in an attempt to fix this cmake error:

```
Imported target "RobotRaconteurCore" includes non-existent path

    "/usr/lib/arm-linux-gnueabihf/dbus-1.0/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.
```

I found the cmake files, and changed strings that were like `/usr/lib/arm-linux-gnueabihf/dbus-1.0/include` to `<sysroot_path>/usr/lib/arm-linux-gnueabihf/dbus-1.0/include`. It seemed to work but now I am thinking that this may be breaking things in subtle ways so I reverted the changes. What is the proper way to handle this error? Do I need to cross compile RR for the beaglebone so that the path prefixes are set correctly?

**Reply by johnwason** [reply]
I have given up on this type of cross compilation using a sysroot because it was an endless loop of problems, and the resulting binaries had subtle problems like incorrect rpaths. People these days typically use a docker with foreign architecture, or just bite the bullet and use native hardware. I have an m1 mac for arm64, so with virtual machines compilation is not a problem. For armhf I am using a raspberry pi with 8 gb of ram which is still pretty slow, but not impractical. Beaglebones use vanilla debian or ubuntu so it is possible to use standard docker images. Older versions of the raspberry pi used armv6 so it requires a custom raspbian image which is very annoying. The arm64 versions of raspberry pi are vanilla debian or ubuntu so a standard arm64 image in parallels on mac works fine.

Modern versions of the Linux kernel support binfmt_misc which transparently allow running armhf and arm64 binaries on an x64 machine. This is what allows for running foreign docker architectures.

I can give an example of cross compiling using docker if you need one, but in general it is worth investing time to understand how to use docker if you do a lot of cross compiling because it saves a huge amount of time.

**Reply by itsLydt** [reply]
Ah okay. Yeah I agree using docker seems like the best course if cross compiling against a sysroot isn't practical. If you have a example that would be great!
